swift/extractor/invocation/SwiftInvocationExtractor.cpp

```diff
@@ -1,7 +1,166 @@
+  ret /= name;
+  ret += '_';
+  ret += hash;
+  ret /= "module";
+  return ret;
+}
+
+std::optional<std::string> getModuleHash(const fs::path& moduleFile) {
+  return getHashOfRealFile(moduleFile);
+}
+
+std::optional<std::string> getModuleHash(const swift::ModuleDecl* module) {
+  return getModuleHash(std::string_view{module->getModuleFilename()});
+}
+
+std::string getSourceId(const fs::path& file) {
+  return "source:"s + file.c_str();
+}
+
+std::string getSourceId(const swift::InputFile& input) {
+  return getSourceId(resolvePath(input.getFileName()));
+}
+
+fs::path getSourceTarget(const fs::path& file) {
+  return "sources" / file.relative_path();
+}
+
+struct ModuleInfo {
+  fs::path target;
+  std::string id;
+};
+
+std::vector<ModuleInfo> emitModuleImplementations(SwiftExtractorState& state,
+                                                  const std::string& moduleName) {
+  std::vector<ModuleInfo> ret;
+  ret.reserve(state.originalOutputModules.size());
+  for (const auto& modulePath : state.originalOutputModules) {
+    if (auto hash = getModuleHash(modulePath)) {
+      auto target = getModuleTarget(moduleName, *hash);
+      if (auto moduleTrap = createTargetTrapDomain(state, target, TrapType::linkage)) {
+        moduleTrap->createLabelWithImplementationId<ModuleDeclTag>(*hash, moduleName);
+        ret.push_back({target, getModuleId(moduleName, *hash)});
+      }
+    }
+  }
+  return ret;
+}
+
+void emitLinkFile(const SwiftExtractorState& state, const fs::path& target, const std::string& id) {
+  if (auto link = createTargetLinkDomain(state, target)) {
+    link->emitTarget(id);
+    link->emitObjectDependency(id);
+  }
+}
+
+void emitSourceObjectDependencies(const SwiftExtractorState& state,
+                                  const fs::path& target,
+                                  const std::string& id) {
+  if (auto object = createTargetObjectDomain(state, target)) {
+    object->emitObject(id);
+    for (auto encounteredModule : state.encounteredModules) {
+      if (auto depHash = getModuleHash(encounteredModule)) {
+        object->emitObjectDependency(getModuleId(encounteredModule, *depHash));
+      }
+    }
+    for (const auto& requestedTrap : state.traps) {
+      object->emitTrapDependency(requestedTrap);
+    }
+  }
+}
+
+void replaceMergedModulesImplementation(const SwiftExtractorState& state,
+                                        const std::string& name,
+                                        const fs::path& mergeTarget,
+                                        const std::string& mergedPartHash) {
+  std::error_code ec;
+  auto mergedPartTarget = getModuleTarget(name, mergedPartHash);
+  fs::copy(getTrapPath(state, mergeTarget, TrapType::linkage),
+           getTrapPath(state, mergedPartTarget, TrapType::linkage),
+           fs::copy_options::overwrite_existing, ec);
+  if (ec) {
+    std::cerr << "unable to replace merged module trap implementation id (" << ec.message() << ")";
```

Comment:
Nit: maybe add the name as well? Otherwise it would be hard to find what exactly went wrong.